src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,342 @@ pub mod policy;
+                        // see `pathlen::self-issued-certs-pathlen` from x509-limbo
+                        // for the testcase we intentionally fail.
+                        //
+                        // Implementation note for someone looking to change this in the future:
+                        // care should be taken to avoid infinite recursion with self-signed
+                        // certificates in the intermediate set; changing this behavior will
+                        // also require a "is not self-signed" check on intermediate candidates.
+                        //
+                        // See https://gist.github.com/woodruffw/776153088e0df3fc2f0675c5e835f7b8
+                        // for an example of this change.
+                        current_depth.checked_add(1).ok_or_else(|| {
+                            ValidationError::Other(
+                                "current depth calculation overflowed".to_string(),
+                            )
+                        })?,
+                        &issuer_extensions,
+                        accumulated_constraints,
+                    ) {
+                        Ok(mut chain) => {
+                            chain.insert(0, working_cert.clone());
+                            return Ok(chain);
+                        }
+                        Err(e) => last_err = Some(e),
+                    };
+                }
+                Err(e) => last_err = Some(e),
+            };
+        }
+
+        // We only reach this if we fail to hit our base case above, or if
+        // a chain building step fails to find a next valid certificate.
+        Err(ValidationError::CandidatesExhausted(last_err.map_or_else(
+            || {
+                Box::new(ValidationError::Other(
+                    "all candidates exhausted with no interior errors".to_string(),
+                ))
+            },
+            |e| match e {
+                // Avoid spamming the user with nested `CandidatesExhausted` errors.
+                ValidationError::CandidatesExhausted(e) => e,
+                _ => Box::new(e),
+            },
+        )))
+    }
+
+    fn build_chain(&self, leaf: &'a Certificate<'chain>) -> Result<Chain<'chain>, ValidationError> {
+        // Before anything else, check whether the given leaf cert
+        // is well-formed according to our policy (and its underlying
+        // certificate profile).
+        //
+        // In the case that the leaf is an EE, this includes a check
+        // against the EE cert's SANs.
+        let leaf_extensions = leaf.extensions()?;
+
+        self.policy.permits_leaf(leaf, &leaf_extensions)?;
+
+        let result = self.build_chain_inner(
+            leaf,
+            0,
+            &leaf_extensions,
+            &mut AccumulatedNameConstraints::default(),
+        );
+        match result {
+            Ok(chain) => Ok(chain),
+            Err(error) => Err(error),
+        }
```

Comment:
No-op, you can just return the value :-)

Reply:
Done!